service/worker/deletenamespace/activities.go

```diff
@@ -0,0 +1,161 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deletenamespace
+
+import (
+	"context"
+
+	enumspb "go.temporal.io/api/enums/v1"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+)
+
+type (
+	activities struct {
+		visibilityManager manager.VisibilityManager
+		metadataManager   persistence.MetadataManager
+		metricsClient     metrics.Client
+		logger            log.Logger
+	}
+)
+
+func NewActivities(
+	visibilityManager manager.VisibilityManager,
+	metadataManager persistence.MetadataManager,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) *activities {
+	return &activities{
+		visibilityManager: visibilityManager,
+		metadataManager:   metadataManager,
+		metricsClient:     metricsClient,
+		logger:            logger,
+	}
+}
+
+func (a *activities) GetNamespaceIDActivity(_ context.Context, nsName namespace.Name) (namespace.ID, error) {
+	getNamespaceRequest := &persistence.GetNamespaceRequest{
+		Name: nsName.String(),
+	}
+
+	getNamespaceResponse, err := a.metadataManager.GetNamespace(getNamespaceRequest)
+	if err != nil {
+		return namespace.EmptyID, err
+	}
+
+	if getNamespaceResponse.Namespace == nil || getNamespaceResponse.Namespace.Info == nil || getNamespaceResponse.Namespace.Info.Id == "" {
+		return namespace.EmptyID, nil
+	}
+
+	return namespace.ID(getNamespaceResponse.Namespace.Info.Id), nil
+}
+
+func (a *activities) MarkNamespaceDeletedActivity(_ context.Context, nsName namespace.Name) error {
+	getNamespaceRequest := &persistence.GetNamespaceRequest{
+		Name: nsName.String(),
+	}
+
+	metadata, err := a.metadataManager.GetMetadata()
+	if err != nil {
+		a.metricsClient.IncCounter(metrics.DeleteNamespaceWorkflowScope, metrics.DeleteNamespaceFailuresCount)
+		a.logger.Error("Unable to get cluster metadata.", tag.WorkflowNamespace(nsName.String()), tag.Error(err))
+		return err
+	}
+
+	ns, err := a.metadataManager.GetNamespace(getNamespaceRequest)
+	if err != nil {
+		a.metricsClient.IncCounter(metrics.DeleteNamespaceWorkflowScope, metrics.DeleteNamespaceFailuresCount)
+		a.logger.Error("Unable to get namespace details.", tag.WorkflowNamespace(nsName.String()), tag.Error(err))
+		return err
+	}
+
+	ns.Namespace.Info.State = enumspb.NAMESPACE_STATE_DELETED
+
+	updateRequest := &persistence.UpdateNamespaceRequest{
+		Namespace:           ns.Namespace,
+		IsGlobalNamespace:   ns.IsGlobalNamespace,
+		NotificationVersion: metadata.NotificationVersion,
+	}
+
+	err = a.metadataManager.UpdateNamespace(updateRequest)
```

Comment:
Because going through FE (and this was my first implementation) makes workflow non-idempotent. Once you marked namespace as `Deleted` you can't access it thought FE any more, and can't rerun workflow if anything goes wrong.

The only thing changed here is namespace state and it is protected with metadata version, so it should not overwrite any potential concurrent changes.